### ISSUE TITLE
First trials (accuracy~53%)

### ISSUE THREAD

**Issue by ruining99** [ruining99/Gravitational-Wave-CNN#8]
I just uploaded data for 3 training trials (trial2, trial3, and trial4). Please ignore trial1, in which the values for mass are not randomized. Also, trial3 and trial4 only add minor modifications on top of trial2 but did not improve accuracy.

**For trial2:**
**Data structure:** _(file "data-stage1-trial2")_
A total of 5701 data is generated with: 
46% gravitational wave, 9% time inverse of gw, 13% empty gaussian, 33% false signals (23% of same magnitude as gw around 10*-21)
The train and test data are generated in one batch, randomly (so extremely unlikely there are overlapping cases); and then later separated

**Data parameters"_(file "data-stage1-trial2")_
snr: 1000, orientation and distance fixed
mass: uniformly generated while guarantees the whole waveform fits within around 0.8 seconds

**model architecture** _(file "classifier-stage1-trial2")_
- The structure follows Daniel George's model.
- compiling: `model.compile(loss='categorical_crossentropy', optimizer='adam', metrics=['accuracy'])`
-fit: `model.fit(X_train, y_train, epochs=10, batch_size=32, verbose=1)`

**Results**
The model converges very quickly to: loss 0.6903 - accuracy: 0.5388
The final evaluation on the test set results in an accuracy of 0.5389

_Trial3 and trial4 attempt to improve the data with modifications, but the accuracy did not improve._
**Trial3:** 11762 data with 56% containing gw. 
The data only contains gw, inverse gw, gaussian noise, and excludes false signals. 
**Results** loss: 0.6833 - acc: 0.5712
accuracy on test set: 0.5631

**Trial4:** 
Removed inverse gw as well so data only contains gaussian and gw.
Tried adding a fourth convolutional layer. 
**Accuracy still remained around 0.5688.**

**Comment by ruining99** [comment]
The above is a brief report that explains what I did. I am quite confused why the accuracy remains low and I am temporarily running out of ideas to improve the data set or the model. Please let me know if you have any ideas :)

Just a thought, is it be possibly better if I train the regression task first on pure gravitational wave? I believe I read in Daniel George's paper that he trained the regression network first and used the same weights to initiate the classification network, which he found increased the performance.

**Comment by ruining99** [comment]
I have fixed the issue in the model. I checked the output of the model and found that it was predicting the exact same thing percentage for every waveform, indicating the model was not learning. The issue was solved by scaling the input data using the sckitlearn function MinMaxScalar. I know we discussed and decided not to scale the data earlier on but after trying many other modifications, only scaling enabled the model to learn. For the first trial I ran (data only contains gaussian and gw), the accuracy reached 100% for both train and test data. I am currently training the model for a more varied dataset. 

I looked up online and found many people facing similar issues when training models. The following are some of the modifications I tried that did not help:
1. randomly initialize the weights
2. change optimiser adam to sgd
3. use sigmoid instead of softmax
4. combine relu in layers
5. simpler model
6. more complex model
7. reduce batch size, lower learning rate,  more epoch, no stratify

Scaling the data may be an issue for the regression network I will be doing next. However,in that case, I guess I can just add a feature for the amplitude of the waveform?
@nathan-jm @chara1ampos

**Comment by nathan-jm** [comment]
I'm very glad to hear that you now have things working in this simple case. Could you add the code to Git, so we can see exactly how the scaling is performed? I presume that you're currently scaling each input by a different amount, so I wonder if it might be possible to scale everything by the same factor and still have it learn well. Or have you already tried this?

You'll indeed have a feature related to the amplitude of the waveform in the regression case, though the physical parameter you want to estimate which scales the waveform is the distance.

**Comment by ruining99** [comment]
I just added the code for the two trials. For trial A, I only used waveform containing gw and waveform of gaussian noise so the net quickly merged to 1.00 accuracy. For trial B, I included the false signals and inverse gravitational waveform. It also converged to 1.00 accuracy after a few epochs. 

The code for the two trials are repetitive so I think you can focus on the one for trial B.

For scaling, I normalized the waveform to be between 0-1 using MinMaxScalar function from Scikitlearn so I am scaling each by a different amount. I have not tried scaling it all by the same amount. I can try that later today. 

If you confirm the code works fine, I can start training on lower signal-noise ratio data and partial waveforms later today as well. @nathan-jm

**Comment by nathan-jm** [comment]
Thanks! The code looks good, from a brief look through it, so you can indeed go ahead with further training.

Thinking about things a bit, I wonder if the problem with training the network without any scaling is because the way the weights are initialized (with kernel_initializer='random_uniform', so they're between -0.05 and 0.05, if I'm reading the documentation correctly) means that the network starts so far away from the optimal weights that it doesn't find the optimum easily. If this is indeed the case, then scaling the input data by some number around 1e21 (so that its values are of order unity, instead of very small) might be all that is necessary, though one might also need to increase the range of initialization of the initial weights.

Of course, the current scaling with MinMaxScaler is fine (since one can just apply the inverse scaling to the output of the network in the regression case), it just seems cleaner to have a uniform scaling applied to all data.

**Comment by ruining99** [comment]
Thank you for the quick feedback! 

In fact, the initialization was an attempt to fix the issue because I saw some sites recommended it. The first trials did not have the initialization parameter (so it would be the default florot uniform initializer) but the issue was still there.

I will still try applying a uniform scaling, maybe around 10^21, to see if it works.
